src/agent/agent.c

```diff
@@ -30,6 +31,65 @@ static bool
                                 void *userdata,
                                 free_func_t free_userdata);
 
+#define AGENT_HEARTBEAT_INTERVAL_USEC (1000000)
+
+static int agent_reset_heartbeat_timer(Agent *agent, sd_event_source **event_source);
+
+static int agent_heartbeat_timer_callback(
+                        sd_event_source *event_source,
+                        UNUSED uint64_t usec,
+                        void *userdata) {
+        Agent *agent = userdata;
+        int r;
+
+        assert(event_source);
+
+        static uint32_t count;
+        printf("%s(): count = %d\n", __func__, count++);
+
+        r = sd_bus_emit_signal(
+                    agent->peer_dbus,
+                    INTERNAL_MANAGER_OBJECT_PATH,
+                    INTERNAL_MANAGER_INTERFACE,
+                    "Heartbeat",
+                    "s",
+                    agent->name);
+        if (r < 0) {
+                fprintf(stderr, "Failed to emit heartbeat signal: %m");
+                return r;
+        }
+
+        r = agent_reset_heartbeat_timer(agent, &event_source);
```

Comment:
Why do you need this. If you sd_event_source_set_enabled to SD_EVENT_ON instead of SD_EVENT_ONESHOT you will automatically get a repeating timeout.

Reply:
From the man page for ```sd_event_source_set_time```:
```
Note that a timer event set to SD_EVENT_ON will fire continuously unless its
configured time is updated using sd_event_source_set_time().
```

I was seeing the timer callback invoked constantly, and I found in some systemd
code that it used ```SD_EVENT_ONESHOT``` and had a "reset" function that calls
```sd_event_source_set_time()```.